# Part 1. Equation of a Slime

How many late days are you using for this assignment? 2

In [30]:
# Imports section
import pandas as pd
import numpy as np
from sklearn.model_selection import cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [20]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)

df = pd.read_csv('science_data_large.csv')

# Output the first 15 rows of the data
print(df.head(15))

# Display a summary of the table information
print(df.info())

    Temperature °C  Mols KCL     Size nm^3
0              469       647  6.244743e+05
1              403       694  5.779610e+05
2              302       975  6.196847e+05
3              779       916  1.460449e+06
4              901        18  4.325726e+04
5              545       637  7.124634e+05
6              660       519  7.006960e+05
7              143       869  2.718260e+05
8               89       461  8.919803e+04
9              294       776  4.770210e+05
10             991       117  2.441771e+05
11             307       781  5.006455e+05
12             206        70  3.145200e+04
13             437       599  5.390215e+05
14             566        75  9.185271e+04
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 n

## 2. Splitting the dataset

In [21]:
# Take the pandas dataset and split it into our features (X) and label (y)

# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X = df[['Temperature °C', 'Mols KCL']]  # Features: Temperature and Mols KCL
y = df['Size nm^3']  # Label: Size nm^3

# Split into training (90%) and test (10%) sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=42)

# Output the sizes of the datasets
print(f"Training set size: {X_train.shape}, {y_train.shape}")
print(f"Test set size: {X_test.shape}, {y_test.shape}")

Training set size: (900, 2), (900,)
Test set size: (100, 2), (100,)


## 3. Perform a Linear Regression

In [22]:
# Use sklearn to train a model on the training set

# Create a sample datapoint and predict the output of that sample with the trained model

# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means

# Extract the coefficients and intercept from the model and write an equation for your h(x) using LaTeX

model = LinearRegression()

# Train the model on the training data
model.fit(X_train, y_train)

# Create a sample datapoint (for example, temperature=300, Mols KCL=600)
sample = [[300, 600]]

# Predict the output for the sample datapoint
predicted_output = model.predict(sample)

# Report the score (R-squared value) for the model
score = model.score(X_test, y_test)

# Output the predicted value and the model score
print(f"Predicted Output for sample {sample}: {predicted_output[0]}")
print(f"Model Score (R²): {score}")

# Extract the coefficients and intercept to form the equation of the model
coefficients = model.coef_
intercept = model.intercept_

# Output the equation of the model
print(f"Equation of the model: h(x) = {intercept} + ({coefficients[0]} * Temperature) + ({coefficients[1]} * Mols)")

Predicted Output for sample [[300, 600]]: 470069.4843196464
Model Score (R²): 0.8552472077276095
Equation of the model: h(x) = -409391.47958340764 + (866.1464133719209 * Temperature) + (1032.6950664857964 * Mols)


/home/codespace/.local/lib/python3.12/site-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


Write the linear equation of a slime: (example equation: $E = mc^2$)

h(x) = -409391.47958340764 + (866.1464133719209 * Temperature) + (1032.6950664857964 * Mols)

Report on score and explain meaning: Conclusion: With an R² score of 0.855, the model is able to explain most of the variance in the size based on temperature and Mols KCL, showing that it is a reasonably strong predictor.

## 4. Use Cross Validation

In [23]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42

# Report on their finding and their significance

cv_scores = cross_val_score(model, X, y, cv=5, scoring='r2')

# Output the cross-validation scores
print("Cross-validation scores:", cv_scores)
print("Mean cross-validation score:", cv_scores.mean())
print("Standard deviation of cross-validation scores:", cv_scores.std())

Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568167899144437
Standard deviation of cross-validation scores: 0.01346630737209602


Write findings here:
Cross-validation scores: [0.83918826 0.87051239 0.85871066 0.87202623 0.84364641]
Mean cross-validation score: 0.8568167899144437
Standard deviation of cross-validation scores: 0.01346630737209602

## 5. Using Polynomial Regression

In [27]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)

# Report on the metrics and output the resultant equation as you did in Part 3.

poly = PolynomialFeatures(degree=2)

# Fit and transform the feature set X
X_poly = poly.fit_transform(X)

# Initialize the model
model_poly = LinearRegression()

# Perform 5-fold cross-validation on the polynomial features
cv_scores_poly = cross_val_score(model_poly, X_poly, y, cv=5, scoring='r2')

# Output the cross-validation scores
print("Cross-validation scores for polynomial features:", cv_scores_poly)
print("Mean cross-validation score:", cv_scores_poly.mean())
print("Standard deviation of cross-validation scores:", cv_scores_poly.std())

# Fit the model on the entire polynomial feature set
model_poly.fit(X_poly, y)

# Get the coefficients and intercept
coefficients_poly = model_poly.coef_
intercept_poly = model_poly.intercept_

# Output the equation
# We will create the equation for h(x), showing each polynomial term and the corresponding coefficient
equation = f"h(x) = {intercept_poly:.4f}"

# Append the terms of the equation
feature_names = poly.get_feature_names_out()
for i, coef in enumerate(coefficients_poly):
    if i == 0:
        continue  # Skip the bias term
    equation += f" + ({coef:.4f} * {feature_names[i]})"

print("\nResultant equation for h(x):\n", equation)

Cross-validation scores for polynomial features: [1. 1. 1. 1. 1.]
Mean cross-validation score: 1.0
Standard deviation of cross-validation scores: 0.0

Resultant equation for h(x):
 h(x) = 0.0000 + (12.0000 * Temperature °C) + (-0.0000 * Mols KCL) + (-0.0000 * Temperature °C^2) + (2.0000 * Temperature °C Mols KCL) + (0.0286 * Mols KCL^2)


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

h(x) = 0.0000 + (12.0000 * Temperature °C) + (-0.0000 * Mols KCL) + (-0.0000 * Temperature °C^2) + (2.0000 * Temperature °C Mols KCL) + (0.0286 * Mols KCL^2)

Report on the score and interpret: While an R² of 1.0 seems ideal, in real-world applications, such a score is highly unrealistic and suggests possible issues with data preparation, model complexity, or data leakage. To validate the model, it should be tested on truly unseen data to confirm whether it generalizes well beyond the training dataset.

# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [31]:
# Load the dataset. Then train and evaluate the classification models.

df = pd.read_csv("ckd_feature_subset.csv")

# Select relevant features and target column
features = ["age", "bp", "wbcc", "appet_poor", "appet_good", "rbcc"]
target = "Target_ckd"

X = df[features]
y = df[target]

# Normalize the features (important for SVM, KNN, Neural Networks)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X)

# Define models
models = {
    "Logistic Regression": LogisticRegression(random_state=42),
    "Support Vector Machine": SVC(random_state=42),
    "k-Nearest Neighbors": KNeighborsClassifier(n_neighbors=5),
    "Neural Network": MLPClassifier(hidden_layer_sizes=(50,), max_iter=500, random_state=42)
}

# Perform 5-fold cross-validation
results = {}
for name, model in models.items():
    scores = cross_val_score(model, X_scaled, y, cv=5, scoring="accuracy")
    results[name] = {"Mean Accuracy": scores.mean(), "Std Dev": scores.std()}

# Convert results to a DataFrame and display
results_df = pd.DataFrame(results).T
print("\nClassification Model Performance (5-Fold Cross-Validation):")
print(results_df)


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(



Classification Model Performance (5-Fold Cross-Validation):
                        Mean Accuracy   Std Dev
Logistic Regression          0.941290  0.024040
Support Vector Machine       0.954194  0.033919
k-Nearest Neighbors          0.941075  0.052810
Neural Network               0.973763  0.024835


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Classification Experiments

In [32]:
# Experiments with Neural Network.

# Neural Network configurations to experiment with
configurations = {
    "Default NN": MLPClassifier(hidden_layer_sizes=(50,), max_iter=500, random_state=42),
    "Larger NN (2 Layers, 100 units)": MLPClassifier(hidden_layer_sizes=(100, 100), max_iter=500, random_state=42),
    "Smaller NN (1 Layer, 10 units)": MLPClassifier(hidden_layer_sizes=(10,), max_iter=500, random_state=42),
    "NN with Tanh Activation": MLPClassifier(hidden_layer_sizes=(50,), max_iter=500, activation='tanh', random_state=42),
}

# Results dictionary to store performance for each configuration
nn_results = {}

# Evaluate each configuration using 5-fold cross-validation
for config_name, model in configurations.items():
    scores = cross_val_score(model, X_scaled, y, cv=5, scoring="accuracy")
    nn_results[config_name] = {"Mean Accuracy": scores.mean(), "Std Dev": scores.std()}

# Convert results to a DataFrame and display
nn_results_df = pd.DataFrame(nn_results).T
print("\nNeural Network Configurations Performance (5-Fold Cross-Validation):")
print(nn_results_df)


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.1


Neural Network Configurations Performance (5-Fold Cross-Validation):
                                 Mean Accuracy   Std Dev
Default NN                            0.973763  0.024835
Larger NN (2 Layers, 100 units)       0.967312  0.029342
Smaller NN (1 Layer, 10 units)        0.947742  0.044437
NN with Tanh Activation               0.960860  0.031967


/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(
/home/codespace/.local/lib/python3.12/site-packages/sklearn/neural_network/_multilayer_perceptron.py:691: ConvergenceWarning: Stochastic Optimizer: Maximum iterations (500) reached and the optimization hasn't converged yet.
  warnings.warn(


## Results and Conclusion for Neural Network Experiments

Neural Network Configurations Performance (5-Fold Cross-Validation):
                                 Mean Accuracy   Std Dev
Default NN                            0.973763  0.024835
Larger NN (2 Layers, 100 units)       0.967312  0.029342
Smaller NN (1 Layer, 10 units)        0.947742  0.044437
NN with Tanh Activation               0.960860  0.031967